url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw5/P3bMain.java

```diff
@@ -0,0 +1,233 @@
+			System.out.print("percentage of errors:");
+			errors = input.nextInt();
+			System.out.print("original filename:");
+			file = input.next();
+		}
+		
+		
+		System.out.println("reading in file...may take a little while");
+		Scanner sc = null;
+		try
+		{
+			sc = new Scanner(new File(file));
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		}
+		
+		String tinput = "";
+		while (sc.hasNextLine())
+		{
+			tinput = tinput+"\n"+sc.nextLine();
+		}
+		
+		String nfile = file.split("\\.")[0];
+		
+		String with_errors = makeErrors(tinput, nfile + "_"+errors + "perror.txt", errors*0.01d);
+		
+		System.out.println("error rate check:");
+		System.out.println("errors: " + compareErrors(tinput, with_errors));
+
+		
+		
+			
+	}
+	
+	
+	/**
+	 * does a spell check
+	 * @param input
+	 */
+	private static void spellcheck(Scanner input)
+	{
+		
+		String tfile = trainingFile;
+		String efile = errorFile;
+		
+		System.out.println("enter file (1) or defaults (2): ");
+		int choice = input.nextInt();
+		
+		if (choice == 1)
+		{
+			System.out.print("file with errors:");
+			efile = input.next();
+			System.out.print("training file without errors:");
+			tfile = input.next();
+		}
+
+		
+		System.out.println("reading in training data...may take a little while");
+		Scanner sc = null;
+		try
+		{
+			sc = new Scanner(new File(tfile));
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		}
+		
+		String tinput = "";
+		while (sc.hasNextLine())
+		{
+			tinput = tinput+"\n"+sc.nextLine();
+		}
+		
+		
+		System.out.println("reading in error text...may take a little while");
+		try
+		{
+			sc = new Scanner(new File(efile));
+		} catch (FileNotFoundException e) {
+			e.printStackTrace();
+		}
+		
+		String einput = "";
+		while (sc.hasNextLine())
+		{
+			einput = einput+"\n"+sc.nextLine();
+		}
+
+		
+		SpellCheck sp = new SpellCheck(tinput, einput);
+		sp.viterbi();
+		
+
+	}
+				
+		
+		
+	
+	/**
+	 * takes an input string and a filename introduces a bunch of errors
+	 * returns the string and also saves the file
+	 * @param input
+	 * @param filename
+	 * @return
+	 */
+	private static String makeErrors(String input, String filename, double probability)
+	{
+		final char[] dict = "abcdefghijklmnopqrstuvwxyz".toCharArray();
+		Random rand = new Random();
+		double dart;
+		String forstring=""; //for returning
+		String forfile = ""; //for writing to file
+		for (char c : input.toCharArray())
+		{
+			dart = rand.nextDouble();
+			if (dart < probability)
+			{
+				if ("abcdefghijklmnopqrstuvwxyz".contains(""+c))
+				{
+					int ra = rand.nextInt(25);
+					while (dict[ra] == c)
+						ra = rand.nextInt(25);
+					c = dict[ra];
+				}
+			}
+			forfile = forfile + c;
+			if (c == '\n')   // clean out carriage returns for string version
+				c = ' ';
+			forstring = forstring + c;
+		}
+			
+		//TODO: do I need to be writing this out to file?
+		try
+		{
+			BufferedWriter out = new BufferedWriter(new FileWriter(filename));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 198, 199. The resource is closed at line 200. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).